swarm_dial.go

```diff
@@ -281,162 +280,375 @@ func (s *Swarm) dialPeer(ctx context.Context, p peer.ID) (*Conn, error) {
 	return nil, err
 }
 
-// doDial is an ugly shim method to retain all the logging and backoff logic
-// of the old dialsync code
-func (s *Swarm) doDial(ctx context.Context, p peer.ID) (*Conn, error) {
-	// Short circuit.
-	// By the time we take the dial lock, we may already *have* a connection
-	// to the peer.
-	c := s.bestAcceptableConnToPeer(ctx, p)
-	if c != nil {
-		return c, nil
-	}
-
-	logdial := lgbl.Dial("swarm", s.LocalPeer(), p, nil, nil)
+///////////////////////////////////////////////////////////////////////////////////
+// lo and behold, The Dialer
+// TODO explain how all this works
+//////////////////////////////////////////////////////////////////////////////////
 
-	// ok, we have been charged to dial! let's do it.
-	// if it succeeds, dial will add the conn to the swarm itself.
-	defer log.EventBegin(ctx, "swarmDialAttemptStart", logdial).Done()
+type dialRequest struct {
+	ctx   context.Context
+	resch chan dialResponse
+}
 
-	conn, err := s.dial(ctx, p)
-	if err != nil {
-		conn := s.bestAcceptableConnToPeer(ctx, p)
-		if conn != nil {
-			// Hm? What error?
-			// Could have canceled the dial because we received a
-			// connection or some other random reason.
-			// Just ignore the error and return the connection.
-			log.Debugf("ignoring dial error because we already have a connection: %s", err)
-			return conn, nil
-		}
+type dialResponse struct {
+	conn *Conn
+	err  error
+}
 
-		// ok, we failed.
-		return nil, err
+// dialWorker is an active dial goroutine that synchronizes and executes concurrent dials
+func (s *Swarm) dialWorker(ctx context.Context, p peer.ID, reqch <-chan dialRequest) error {
```

Comment:
uber ultra nit: `startDialWorker` (because it goes async).

Reply:
agreed, not so nitty :) I will rename.

Reply:
done.